File: src/infrastructure/persistence/file_knowledge_reader.py
"""File-based knowledge reader adapter.

FileKnowledgeReader implements KnowledgeReaderProtocol by reading
knowledge entries from the filesystem.  TOML files are parsed for
``[_meta]`` sections to extract EntryMetadata.  Markdown files are
returned with their raw content.
"""

from __future__ import annotations

import tomllib
from typing import TYPE_CHECKING

from src.domain.models.errors import InvariantViolationError

if TYPE_CHECKING:
    from pathlib import Path
from src.domain.models.knowledge_entry import (
    EntryMetadata,
    KnowledgeCategory,
    KnowledgeEntry,
    KnowledgePath,
)


class FileKnowledgeReader:
    """Reads knowledge entries from a directory on the local filesystem.

    Directory layout::

        {knowledge_dir}/
        ├── ddd/{topic}.md
        ├── conventions/{topic}.md
        ├── tools/{tool}/{version}/{topic}.toml
        └── cross-tool/{topic}.toml
    """

    def __init__(self, knowledge_dir: Path) -> None:
        self._root = knowledge_dir

    def read_entry(
        self,
        path: KnowledgePath,
        version: str = "current",
    ) -> KnowledgeEntry:
        """Read a knowledge entry from the filesystem.

        Args:
            path: The RLM-addressable knowledge path.
            version: Version directory for tools entries (default "current").

        Returns:
            A KnowledgeEntry populated from the file.

        Raises:
            InvariantViolationError: If the file does not exist.
        """
        file_path = self._resolve_file_path(path, version)
        if not file_path.exists():
            msg = f"Knowledge entry not found: {path.raw} (looked at {file_path})"
            raise InvariantViolationError(msg)

        if file_path.suffix == ".toml":
            return self._read_toml_entry(path, file_path)
        return self._read_markdown_entry(path, file_path)

    def list_topics(
        self,
        category: KnowledgeCategory,
        tool: str | None = None,
    ) -> tuple[str, ...]:
        """List available topics within a category.

        Args:
            category: The knowledge category to scan.
            tool: For TOOLS category, the specific tool to list topics for.

        Returns:
            Sorted tuple of topic stem names.
        """
        if category == KnowledgeCategory.TOOLS:
            if tool is None:
                return ()
            scan_dir = self._root / "tools" / tool / "current"
            pattern = "*.toml"
        elif category == KnowledgeCategory.CROSS_TOOL:
            scan_dir = self._root / "cross-tool"
            pattern = "*.toml"
        elif category == KnowledgeCategory.DDD:
            scan_dir = self._root / "ddd"
            pattern = "*.md"
        else:  # CONVENTIONS
            scan_dir = self._root / "conventions"
            pattern = "*.md"

        if not scan_dir.exists():
            return ()

        return tuple(sorted(f.stem for f in scan_dir.glob(pattern)))

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _resolve_file_path(self, path: KnowledgePath, version: str) -> Path:
        """Map a KnowledgePath to a concrete filesystem path."""
        if path.category == KnowledgeCategory.TOOLS:
            tool = path.tool
            subtopic = path.subtopic
            return self._root / "tools" / (tool or "") / version / f"{subtopic}.toml"

        if path.category == KnowledgeCategory.CROSS_TOOL:
            topic = path.topic
            return self._root / "cross-tool" / f"{topic}.toml"

        # DDD and CONVENTIONS: {category}/{topic}.md
        category_dir: str = path.category.value
        return self._root / category_dir / f"{path.topic}.md"

    def _read_toml_entry(
        self,
        path: KnowledgePath,
        file_path: Path,
    ) -> KnowledgeEntry:
        """Parse a TOML file into a KnowledgeEntry."""
        raw_text = file_path.read_text()
        data = tomllib.loads(raw_text)
        metadata = self._extract_metadata(data)
        return KnowledgeEntry(
            path=path,
            title=path.subtopic or path.topic,
            content=raw_text,
            metadata=metadata,
            format="toml",
        )

    def _read_markdown_entry(
        self,
        path: KnowledgePath,
        file_path: Path,
    ) -> KnowledgeEntry:
        """Read a Markdown file into a KnowledgeEntry."""
        content = file_path.read_text()
        return KnowledgeEntry(
            path=path,
            title=path.topic,
            content=content,
            metadata=None,
            format="markdown",
        )

    @staticmethod
    def _extract_metadata(data: dict[str, object]) -> EntryMetadata | None:
        """Extract EntryMetadata from a parsed TOML ``[_meta]`` section."""
        meta_raw = data.get("_meta")
        if not isinstance(meta_raw, dict):
            return None

        source_urls_raw = meta_raw.get("source_urls", ())
        source_urls: tuple[str, ...] = ()
        if isinstance(source_urls_raw, list):
            source_urls = tuple(str(u) for u in source_urls_raw)

        schema_raw = meta_raw.get("schema_version")
        schema_version = str(schema_raw) if schema_raw is not None else None

        return EntryMetadata(
            last_verified=_str_or_none(meta_raw.get("last_verified")),
            verified_against=_str_or_none(meta_raw.get("verified_against")),
            confidence=str(meta_raw.get("confidence", "high")),
            deprecated=bool(meta_raw.get("deprecated", False)),
            next_review_date=_str_or_none(meta_raw.get("next_review_date")),
            schema_version=schema_version,
            source_urls=source_urls,
        )


def _str_or_none(val: object) -> str | None:
    """Convert a value to str or None."""
    if val is None:
        return None
    return str(val)

File: internal/knowledge/infrastructure/file_knowledge_reader.go
// Package infrastructure provides adapters for the Knowledge bounded context.
package infrastructure

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/BurntSushi/toml"

	knowledgedomain "github.com/alty-cli/alty/internal/knowledge/domain"
	domainerrors "github.com/alty-cli/alty/internal/shared/domain/errors"
)

// FileKnowledgeReader reads knowledge entries from the local filesystem.
//
// Directory layout:
//
//	{knowledge_dir}/
//	  ddd/{topic}.md
//	  conventions/{topic}.md
//	  tools/{tool}/{version}/{topic}.toml
//	  cross-tool/{topic}.toml
type FileKnowledgeReader struct {
	root string
}

// NewFileKnowledgeReader creates a FileKnowledgeReader.
func NewFileKnowledgeReader(knowledgeDir string) *FileKnowledgeReader {
	return &FileKnowledgeReader{root: knowledgeDir}
}

// ReadEntry reads a knowledge entry from the filesystem.
func (r *FileKnowledgeReader) ReadEntry(_ context.Context, path knowledgedomain.KnowledgePath, version string) (knowledgedomain.KnowledgeEntry, error) {
	filePath := r.resolveFilePath(path, version)

	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		return knowledgedomain.KnowledgeEntry{}, fmt.Errorf(
			"knowledge entry not found: %s (looked at %s): %w",
			path.Raw(), filePath, domainerrors.ErrInvariantViolation)
	}

	if strings.HasSuffix(filePath, ".toml") {
		return r.readTOMLEntry(path, filePath)
	}
	return r.readMarkdownEntry(path, filePath)
}

// ListTopics lists available topics within a category.
func (r *FileKnowledgeReader) ListTopics(_ context.Context, category knowledgedomain.KnowledgeCategory, tool *string) ([]string, error) {
	var scanDir, pattern string

	switch category {
	case knowledgedomain.CategoryTools:
		if tool == nil {
			return nil, nil
		}
		scanDir = filepath.Join(r.root, "tools", *tool, "current")
		pattern = "*.toml"
	case knowledgedomain.CategoryCrossTool:
		scanDir = filepath.Join(r.root, "cross-tool")
		pattern = "*.toml"
	case knowledgedomain.CategoryDDD:
		scanDir = filepath.Join(r.root, "ddd")
		pattern = "*.md"
	case knowledgedomain.CategoryConventions:
		scanDir = filepath.Join(r.root, "conventions")
		pattern = "*.md"
	}

	if _, err := os.Stat(scanDir); os.IsNotExist(err) {
		return nil, nil
	}

	matches, err := filepath.Glob(filepath.Join(scanDir, pattern))
	if err != nil {
		return nil, fmt.Errorf("globbing %s: %w", scanDir, err)
	}

	topics := make([]string, 0, len(matches))
	for _, m := range matches {
		base := filepath.Base(m)
		ext := filepath.Ext(base)
		topics = append(topics, strings.TrimSuffix(base, ext))
	}
	sort.Strings(topics)
	return topics, nil
}

func (r *FileKnowledgeReader) resolveFilePath(path knowledgedomain.KnowledgePath, version string) string {
	category := path.Category()

	if category == knowledgedomain.CategoryTools {
		tool := path.Tool()
		subtopic := path.Subtopic()
		toolStr := ""
		if tool != nil {
			toolStr = *tool
		}
		subtopicStr := ""
		if subtopic != nil {
			subtopicStr = *subtopic
		}
		return filepath.Join(r.root, "tools", toolStr, version, subtopicStr+".toml")
	}

	if category == knowledgedomain.CategoryCrossTool {
		return filepath.Join(r.root, "cross-tool", path.Topic()+".toml")
	}

	// DDD and conventions: {category}/{topic}.md
	return filepath.Join(r.root, string(category), path.Topic()+".md")
}

func (r *FileKnowledgeReader) readTOMLEntry(path knowledgedomain.KnowledgePath, filePath string) (knowledgedomain.KnowledgeEntry, error) {
	rawText, err := os.ReadFile(filePath)
	if err != nil {
		return knowledgedomain.KnowledgeEntry{}, fmt.Errorf("reading %s: %w", filePath, err)
	}

	var data map[string]any
	if _, err := toml.Decode(string(rawText), &data); err != nil {
		return knowledgedomain.KnowledgeEntry{}, fmt.Errorf("decoding TOML %s: %w", filePath, err)
	}

	metadata := extractMetadata(data)

	title := path.Topic()
	if sub := path.Subtopic(); sub != nil {
		title = *sub
	}

	return knowledgedomain.NewKnowledgeEntry(path, title, string(rawText), metadata, "toml"), nil
}

func (r *FileKnowledgeReader) readMarkdownEntry(path knowledgedomain.KnowledgePath, filePath string) (knowledgedomain.KnowledgeEntry, error) {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return knowledgedomain.KnowledgeEntry{}, fmt.Errorf("reading %s: %w", filePath, err)
	}
	return knowledgedomain.NewKnowledgeEntry(path, path.Topic(), string(content), nil, "markdown"), nil
}

func extractMetadata(data map[string]any) *knowledgedomain.EntryMetadata {
	metaRaw, ok := data["_meta"]
	if !ok {
		return nil
	}
	metaMap, ok := metaRaw.(map[string]any)
	if !ok {
		return nil
	}

	var sourceURLs []string
	if urlsRaw, ok := metaMap["source_urls"]; ok {
		if urlsList, ok := urlsRaw.([]any); ok {
			for _, u := range urlsList {
				sourceURLs = append(sourceURLs, fmt.Sprintf("%v", u))
			}
		}
	}

	schemaVersion := ""
	if sv, ok := metaMap["schema_version"]; ok && sv != nil {
		schemaVersion = fmt.Sprintf("%v", sv)
	}

	confidence := "high"
	if c, ok := metaMap["confidence"]; ok {
		confidence = fmt.Sprintf("%v", c)
	}

	deprecated := false
	if d, ok := metaMap["deprecated"]; ok {
		if b, ok := d.(bool); ok {
			deprecated = b
		}
	}

	meta := knowledgedomain.NewEntryMetadata(
		strOrEmpty(metaMap["last_verified"]),
		strOrEmpty(metaMap["verified_against"]),
		confidence,
		deprecated,
		strOrEmpty(metaMap["next_review_date"]),
		schemaVersion,
		sourceURLs,
	)
	return &meta
}

func strOrEmpty(val any) string {
	if val == nil {
		return ""
	}
	return fmt.Sprintf("%v", val)
}
